File: Data-Structures-And-Algorithms/BinarySearch/leetCode-33-search.go
func search(nums []int, target int) int {
	minIdx := findMinIdx(nums)
	if nums[minIdx] == target {
		return minIdx
	}
	left, right := 0, len(nums)-1
	if target <= nums[right] {
		left = minIdx + 1
	} else {
		right = minIdx - 1
	}
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] == target {
			return mid
		} else if nums[mid] > target {
			right = mid - 1
		} else if nums[mid] < target {
			left = mid + 1
		}
	}
	return -1
}
func findMinIdx(nums []int) int {
	left, right := 0, len(nums)-1
	for left < right {
		mid := left + (right-left)/2
		if nums[mid] > nums[right] {
			left = mid + 1
		} else {
			right = mid
		}
	}
	return left
}

File: Data-Structures-And-Algorithms/BinarySearch/leetCode-33-search.java
class Solution {
    public int search(int[] nums, int target) {
        int n = nums.length;
        if (n <= 0) return -1;
        if (n == 1) return nums[0] == target ? 0 : -1;

        int left = 0, right = n - 1;
        while (left <= right) {
            int mid = left + (right-left)/2;
            if (nums[mid] == target) return mid;
            if (nums[0] <= nums[mid]) {
                if (nums[0] <= target && target < nums[mid]) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[mid] < target && target <= nums[n-1]) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }
        return -1;
    }
}
